clients/drcachesim/drpt2trace/pt2ir.cpp

```diff
@@ -0,0 +1,391 @@
+}
+
+uint64_t
+pt2ir_t::get_instr_count()
+{
+    return pt_insn_list_.size();
+}
+
+void
+pt2ir_t::print_instrs_to_stdout()
+{
+    void *dcontext = GLOBAL_DCONTEXT;
+    for (auto pt_instr : pt_insn_list_) {
+        disassemble_with_info(dcontext, pt_instr.raw, STDOUT, false, false);
+    }
+}
+
+bool
+pt2ir_t::load_pt_raw_file(IN std::string &path)
+{
+    /* Under C++11, there is no good solution to get the file size after using ifstream to
+     * open a file. Because we will not change the PT raw trace file during converting, we
+     * don't need to think about write-after-read. We get the file size from file stat
+     * first and then use ifstream to open and read the PT raw trace file.
+     * XXX: We may need to update Dynamorio to support C++17+. Then we can implement the
+     * following logic without opening the file twice.
+     */
+    errno = 0;
+    struct stat fstat;
+    if (stat(path.c_str(), &fstat) == -1) {
+        ERRMSG("Failed to get file size of trace file: %s: %d.\n", path.c_str(), errno);
+        return false;
+    }
+    pt_raw_buffer_size_ = static_cast<size_t>(fstat.st_size);
+
+    pt_raw_buffer_ = std::unique_ptr<unsigned char>(new uint8_t[pt_raw_buffer_size_]);
+    std::ifstream f(path, std::ios::binary | std::ios::in);
+    if (!f.is_open()) {
+        ERRMSG("Failed to open trace file: %s.\n", path.c_str());
+        return false;
+    }
+    f.read(reinterpret_cast<char *>(pt_raw_buffer_.get()), pt_raw_buffer_size_);
+    if (f.fail()) {
+        ERRMSG("Failed to read trace file: %s.\n", path.c_str());
+        return false;
+    }
+
+    f.close();
+    return true;
+}
+
+bool
+pt2ir_t::load_kernel_image(IN std::string &path)
+{
+    struct pt_image *kimage = pt_sb_kernel_image(pt_sb_session_);
+    /* Load all ELF sections in kcore to the shared image cache.
+     * XXX: load_elf() is implemented in libipt's client ptxed. Currently we directly use
+     * it. We may need to implement a c++ version in our client.
+     */
+    int errcode =
+        load_elf(pt_iscache_, pt_sb_kernel_image(pt_sb_session_), path.c_str(), 0, "", 0);
+    if (errcode < 0) {
+        ERRMSG("Failed to load kernel image %s: %s.\n", path.c_str(),
+               pt_errstr(pt_errcode(errcode)));
+        return false;
+    }
+    return true;
+}
+
+bool
+pt2ir_t::alloc_sb_pevent_decoder(IN struct pt_sb_pevent_config *config)
+{
+    int errcode = pt_sb_alloc_pevent_decoder(pt_sb_session_, config);
+    if (errcode < 0) {
+        ERRMSG("Failed to allocate sideband perf event decoder: %s.\n",
+               pt_errstr(pt_errcode(errcode)));
+        return false;
+    }
+    return true;
+}
+
+void
+pt2ir_t::dx_decoding_error(IN int errcode, IN const char *errtype, IN uint64_t ip)
+{
+    int err = -pte_internal;
+    uint64_t pos = 0;
+
+    /* Get the current position of 'pt_instr_decoder_'. It will fill the position into
+     * pos. The 'pt_insn_get_offset' function mainly used to report error.
```

Comment:
grammar: s/mainly/is mainly/; s/error/errors/

Reply:
Done.